jupyterhub/apihandlers/proxy.py

```diff
@@ -9,28 +9,49 @@
 from tornado import gen, web
 
 from .. import orm
+from ..proxy import RouteSpec
 from ..utils import admin_only
 from .base import APIHandler
 
+class _RouteSpecJSONEncoder(json.JSONEncoder):
```

Comment:
This is the part I like the least - now that routes are no longer valid keys in a JSON dict, we need to cast them to strings. And since we are casting them to strings, maybe we should just be using strings anyway?

Since paths must always start with `/`, a routespec can already be unambiguous as a URL:

    'host/path/...'
    '/path/only/...'